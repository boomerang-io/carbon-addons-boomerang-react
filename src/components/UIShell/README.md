# UI Shell Component

> React component that includes Header and Sidenav
