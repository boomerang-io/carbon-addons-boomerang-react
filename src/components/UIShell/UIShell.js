import React from 'react';
import PropTypes from 'prop-types';

import Header from '../Header'; // Using default export
import HeaderMenuButton from '../HeaderMenuButton';
import HeaderMenuLink from '../HeaderMenuLink';
import HeaderMenuUser from '../HeaderMenuUser';
import AboutPlatform from '../AboutPlatform';
import ContactUs from '../ContactUs';
import PrivacyStatement from '../PrivacyStatement';
import ReportBug from '../ReportBug';
import SignOut from '../SignOut';
import GdprRedirectModal from '../GdprRedirectModal';

UIShell.propTypes = {
  /**
   * used in header, after the logo or platformName
   */
  appName: PropTypes.string,
  /**
   * Base service url used to build up several requests
   * used by the header. paths to service endpoints are hardcoded
   */
  baseServiceUrl: PropTypes.string,
  /**
   * base launch url, used to redirect to Launchpad if need to consent per GDPR
   */
  baseLaunchEnvUrl: PropTypes.string,

  /**
   * alias for "platformName" for legacy support
   */
  companyName: PropTypes.string,

  /**
   * Pass in whole header config object used for
   * - Feature flagging
   * - Platform links
   * - Metadata about the platform
   * - Message banner to display for all users
   */
  headerConfig: PropTypes.shape({
    features: PropTypes.object,
    navigation: PropTypes.array,
    platform: PropTypes.object,
    platformMessage: PropTypes.object,
  }),
  /**
   * Used to render the hamburger menu icon and the component passed
   * in by the user
   */
  onMenuClick: PropTypes.func,

  /**
   * Used to trigger tutorial or any arbitrary function passed
   * in on click of "Tutorial" header menu item
   */
  onTutorialClick: PropTypes.func,
  /**
   * used in header, only rendered if renderLogo is false
   */
  platformName: PropTypes.string,
  /**
   * alias for "appName" for legacy support
   */
  productName: PropTypes.string,
  /**
   * override for the consumer. When set as true, the launchpad redirect modal will be
   */
  renderLogo: PropTypes.bool,
  requirePlatformConsent: PropTypes.bool,
  /**
   * Pass in whole user object
   */
  user: PropTypes.shape({
    email: PropTypes.string,
    hasConsented: PropTypes.bool,
    name: PropTypes.string,
  }),

  /**
   * Icon that is added by each team for their custom behaviour
   */
  renderRightPanel: PropTypes.shape({
    icon: PropTypes.element,
    component: PropTypes.element,
  }),
  /**
   * alias for "onMenuClick" that is more descriptive
   * defaults to "onMenuClick" if both props are passed
   */
  renderSidenav: PropTypes.func,

  /**
   * Used to navigate to the main content of the app
   */
  skipToContentProps: PropTypes.shape({
    href: PropTypes.string,
    children: PropTypes.string,
    className: PropTypes.string,
  }),
};

UIShell.defaultProps = {
  baseLaunchEnvUrl: '',
  baseWWWEnvUrl: '',
  headerConfig: {},
  requirePlatformConsent: true,
  user: {},
  renderRightPanel: {},
};

function UIShell({
  appName,
  baseLaunchEnvUrl,
  baseServiceUrl,
  companyName,
  headerConfig,
  onMenuClick,
  onTutorialClick,
  platformName,
  productName,
  renderLogo,
  requirePlatformConsent,
  renderRightPanel,
  renderSidenav,
  skipToContentProps,
  user,
}) {
  const { features, navigation, platform, platformMessage } = headerConfig;

  const finalPlatformName = platformName || companyName;
  const finalAppName = appName || productName;
  const isSupportEnabled = Boolean(features) && Boolean(features['support.enabled']);

  return (
    <>
      <Header
        appName={finalAppName}
        baseLaunchEnvUrl={baseLaunchEnvUrl}
        enableNotifications={features && features['notifications.enabled']}
        platformMessage={platformMessage}
        platformName={!renderLogo && finalPlatformName ? finalPlatformName : null}
        renderLogo={renderLogo}
        navLinks={navigation}
        notificationsConfig={{
          wsUrl: `${baseServiceUrl}/notifications/ws`,
        }}
        onHelpClick={[
          typeof onTutorialClick === 'function' && (
            <HeaderMenuButton
              text="Tutorial"
              iconName="workspace"
              onClick={onTutorialClick}
              key="Tutorial"
            />
          ),
          baseServiceUrl && isSupportEnabled && (
            <ContactUs baseServiceUrl={baseServiceUrl} key="Contact Us" />
          ),
          baseServiceUrl && isSupportEnabled && (
            <ReportBug baseServiceUrl={baseServiceUrl} key="Report Bug" />
          ),
          baseServiceUrl && isSupportEnabled && (
            <HeaderMenuLink
              external={false}
              href={`${baseLaunchEnvUrl}/launchpad/support`}
              iconName="support"
              text="Support Center"
            />
          ),
          platform && platform.communityUrl && (
            <HeaderMenuLink href={platform.communityUrl} iconName="forum" text="Community" />
          ),
        ].filter(Boolean)}
        profileChildren={[
          user?.id && (
            <HeaderMenuUser
              key="Avatar"
              src={`${baseServiceUrl}/users/image/${user.email}`}
              userName={user.name}
            />
          ),
          platform && (
            <AboutPlatform
              key="About Platform"
              organization={platform.name}
              version={platform.version}
            />
          ),
          baseServiceUrl && (
            <PrivacyStatement key="Privacy Statement" baseServiceUrl={baseServiceUrl} />
          ),
          platform && platform.signOutUrl && (
            <SignOut signOutLink={platform.signOutUrl} key="Sign Out" />
          ),
        ].filter(Boolean)}
        renderRightPanel={renderRightPanel}
        skipToContentProps={skipToContentProps}
        renderSidenav={onMenuClick || renderSidenav}
      />
      {requirePlatformConsent && user.hasOwnProperty('hasConsented') && !user.hasConsented ? (
        <GdprRedirectModal isOpen baseLaunchEnvUrl={baseLaunchEnvUrl} user={user} />
      ) : null}
    </>
  );
}

export default UIShell;
