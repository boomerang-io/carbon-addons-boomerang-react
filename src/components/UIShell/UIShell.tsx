/*
IBM Confidential
694970X, 69497O0
Â© Copyright IBM Corp. 2022, 2024
*/


import React from "react";
import { QueryClientProvider } from "react-query";
import { Book, Forum, HelpDesk, Email, ChatLaunch, Cognitive } from "@carbon/react/icons";
import Header from "../Header"; // Using default export
import HeaderMenuItem from "../Header/HeaderMenuItem";
import PrivacyRedirectModal from "../PrivacyRedirect";
import { AboutPlatformMenuItem } from "../AboutPlatform";
import { FeedbackMenuItem } from "../Feedback";
import { SupportCenterMenuItem } from "../SupportCenter";
import { PrivacyStatementMenuItem } from "../PrivacyStatement";
import { ProfileSettingsMenuItem } from "../ProfileSettings";
import { queryClient } from "../../config/servicesConfig";
import { SignOutMenuItem } from "../SignOut";
import type { NavLink, User } from "../../types";
import { USER_PLATFORM_ROLE } from "../../constants/UserType";

type Props = {
  carbonTheme?: "white" | "g10" | "g90" | "g100";
  config?: {
    features?: {
      "appSwitcher.enabled"?: boolean;
      "consent.enabled"?: boolean;
      "docs.enabled"?: boolean;
      "eventing.enabled"?: boolean;
      "feedback.enabled"?: boolean;
      "homePage.enabled"?: boolean;
      "partner.enabled"?: boolean;
      "metering.enabled"?: boolean;
      "notifications.enabled"?: boolean;
      "notificationsCount.enabled"?: boolean;
      "support.enabled"?: boolean;
      "welcome.enabled"?: boolean;
    };
    navigation?: NavLink[];
    platform: {
      baseEnvUrl: string;
      baseServicesUrl: string;
      communityUrl?: string;
      feedbackUrl?: string;
      name?: string; // Used in About Plaform modal
      platformEmail?: string;
      platformName: string; // Used in UIShell
      platformOrganization?: string;
      privateTeams?: boolean;
      sendIdeasUrl?: string;
      sendMail?: boolean;
      askICAEnabled?:boolean;
      signOutUrl?: string;
      version?: string;
      assistantVersion?:string;
      agentsVersion?:string;
      scribeFlowVersion?:string;
      askICAUrl?:string;
      supportUrl?:string;
    };
    platformMessage?: any;
  };
  leftPanel?: (args: { close: () => void; isOpen: boolean; navLinks?: NavLink[] }) => React.ReactNode;
  platformName?: string;
  productName?: string;
  profileMenuItems?: React.ReactNode[];
  renderPrivacyRedirect?: boolean;
  renderPrivacyStatement?: boolean;
  rightPanel?: { icon?: React.ReactNode; component: React.ReactNode };
  skipToContentProps?: {
    href?: string;
    children?: string;
    className?: string;
  };
  supportMenuItems?: React.ReactNode[];
  templateMeteringEvent?: (props: any) => void;
  triggerEvent?: (props: any) => any;
  user?: User;
  userTeams?: { data: any, isLoading: boolean, error: any };
  enableIcaMacs?: boolean;
  handleShowTutorial?: Function;
  tutorialScreenToShow?: string;
};


function UIShell({
  carbonTheme="g10",
  config,
  leftPanel,
  platformName,
  productName,
  profileMenuItems = [],
  supportMenuItems = [],
  renderPrivacyRedirect = true,
  renderPrivacyStatement = true,
  rightPanel,
  handleShowTutorial,
  skipToContentProps,
  templateMeteringEvent,
  triggerEvent,
  tutorialScreenToShow,
  user,
  userTeams,
  enableIcaMacs,
}: Props) {
  // Support base header .e.g for an error state
  if (!config) {
    return (
      <Header
        baseEnvUrl=""
        baseServicesUrl=""
        carbonTheme={carbonTheme}
        enableAppSwitcher={false}
        enableNotifications={false}
        enableNotificationsCount={false}
        productName={productName || platformName || ""}
      />
    );
  }
  const { features, navigation, platform, platformMessage } = config;
  const names = getProductAndPlatformNames({ productName, platformName, platform });
  const sendIdeasUrl = platform?.feedbackUrl || "https://ideas.ibm.com";
  const supportLink = platform?.supportUrl;
  const partnerEmailId="ica-support@ibm.com";
  /**
   * Check feature enablement via explicit feature flags
   */
  const isAppSwitcherEnabled = Boolean(features?.["appSwitcher.enabled"]);
  const isFeedbackEnabled = Boolean(features?.["feedback.enabled"]);
  const isNotificationsEnabled = Boolean(features?.["notifications.enabled"]);
  const isNotificationsCountEnabled = Boolean(features?.["notificationsCount.enabled"]);
  const isSupportEnabled = Boolean(features?.["support.enabled"]);
  /**
   * Check feature enablement via value truthiness
   */
  const isAboutPlatformEnabled = Boolean(platform.name && platform.version);
  const isCommunityEnabled = Boolean(platform?.communityUrl);
  const isSendMailEnabled = Boolean(platform.sendMail);
  const isSignOutEnabled = Boolean(platform?.signOutUrl);
  const isUserEnabled = Boolean(user?.id);
  const isPartnerUser = Boolean(user?.type === USER_PLATFORM_ROLE.Partner);
  const supportFlagCheck = user?.showSupport;
  const askICAEnabled=Boolean(platform?.askICAEnabled);

  /**
   * Checking for conditions when we explicitly set "renderPrivacyRedirect" to false (it defaults to true) OR
   * it's disabled overall for the platform. This lets us toggle the UIShell consent redirect per app as needed
   * e.g. disabled in Launchpad, but have it enabled for rest of the platform AND also support
   * having it disabled in a "standalone" mode via the consent.enable feature flag, i.e. data driven via the service
   */
  const isPrivacyRedirectDisabled = renderPrivacyRedirect === false || features?.["consent.enabled"] === false;
  const isPrivacyModalRendered = isPrivacyRedirectDisabled === false && user?.hasConsented === false;

  /**
   * Also enable/disable privacy statement via the consent.enabled feature flag
   */
  const isPrivacyStatementDisabled = renderPrivacyStatement === false || features?.["consent.enabled"] === false;

  return (
    <QueryClientProvider client={queryClient}>
      <Header
        baseEnvUrl={platform.baseEnvUrl}
        baseServicesUrl={platform.baseServicesUrl}
        carbonTheme={carbonTheme}
        enableAppSwitcher={isAppSwitcherEnabled}
        enableNotifications={isNotificationsEnabled}
        enableNotificationsCount={isNotificationsCountEnabled}
        leftPanel={leftPanel}
        navLinks={navigation}
        platformMessage={platformMessage}
        prefixName={names.platformName}
        productName={names.productName}
        rightPanel={rightPanel}
        requestSummary={user?.requestSummary}
        skipToContentProps={skipToContentProps}
        templateMeteringEvent={templateMeteringEvent}
        triggerEvent={triggerEvent}
        profileMenuItems={[
          isUserEnabled && (
            <ProfileSettingsMenuItem
              key="profile-settings"
              baseServicesUrl={platform.baseServicesUrl}
              src={`${platform.baseServicesUrl}/users/image/${user?.email}`}
              userName={user?.displayName ?? user?.name}
            />
          ),
          isSendMailEnabled && (
            <HeaderMenuItem
              key="email-preferences"
              href={`${platform.baseEnvUrl}/launchpad/email-preferences`}
              icon={<Email />}
              kind="internal"
              text="Email Preferences"
              type="link"
            />
          ),
          !isPrivacyStatementDisabled && (
            <PrivacyStatementMenuItem
              key="privacy-statement"
              baseServicesUrl={platform.baseServicesUrl}
              platformEmail={platform?.platformEmail}
            />
          ),
          ...profileMenuItems,
          isSignOutEnabled && <SignOutMenuItem key="Sign Out" signOutLink={platform.signOutUrl as string} />,
        ].filter(Boolean)}
        supportMenuItems={[
          isSupportEnabled && (
            (supportFlagCheck|| isPartnerUser) ?
            <SupportCenterMenuItem
                key="support-center"
                platformName={platform?.platformName}
                platformOrganization={platform?.platformOrganization}
                supportLink={supportLink}
                partnerEmailId={partnerEmailId}
                enablePartner={isPartnerUser}
                baseServicesUrl={platform.baseServicesUrl}
              />
             :
             <HeaderMenuItem
             key="support-center"
             href={platform?.supportUrl as string}
             icon={<HelpDesk />}
             kind="external"
             text="Support Center"
             type="link"
           />   
          ),
          isCommunityEnabled && (
            <HeaderMenuItem
              key="community"
              href={platform?.communityUrl as string}
              icon={<Forum />}
              kind="external"
              text="Community"
              type="link"
            />
          ),
          isFeedbackEnabled && (
            <FeedbackMenuItem
              key="feedback"
              platformName={platform?.platformName}
              platformOrganization={platform?.platformOrganization}
              sendIdeasUrl={sendIdeasUrl}
            />
          ),
          ((!isPartnerUser && askICAEnabled)?(
          <HeaderMenuItem
          key="chat-launch"
          href={platform?.askICAUrl as string}
          icon={< ChatLaunch/>}
          data-testid="askICA-chatlaunch"
          kind="external"
          text="AskICA"
          type="link"
        />):null),
        (!isPartnerUser && enableIcaMacs &&
        <HeaderMenuItem
          key="launch-ideation-agent"
          href={`${platform.baseEnvUrl}/launchpad/macs`}
          icon={<Cognitive/>}
          data-testid="launch-ideation-agent"
          kind="app"
          text="watsonx Challenge Ideation"
          type="link"
        />
        ),
        isAboutPlatformEnabled && (
          <AboutPlatformMenuItem
            key="about-platform"
            name={platform.name as string}
            version={platform.version as string}
            assistantVersion={platform.assistantVersion as string}
            agentsVersion={platform.agentsVersion as string}
            scribeFlowVersion={platform.scribeFlowVersion as string}
          />
        ),
        (tutorialScreenToShow && handleShowTutorial &&
          <HeaderMenuItem
            key="launch-tutorial"
            onClick={handleShowTutorial as () => void}
            icon={<Book />}
            data-testid="launch-tutorial"
            text="Launch Tutorial"
            type="button"
          />
        ),
          ...supportMenuItems,
        ].filter(Boolean)}
        userTeams={userTeams}
      />
      {isPrivacyModalRendered ? (
        <PrivacyRedirectModal
          isOpen
          baseEnvUrl={platform.baseEnvUrl as string}
          platformName={platform?.name}
          user={user}
        />
      ) : null}
    </QueryClientProvider>
  );
}

/**
 * Determine how to render the name and prefix in the Header based
 *  on what is passed in. If we only have the plaform or product name, then
 * we want it to be bolded. If we have both, then make the platform the prefix
 */
function getProductAndPlatformNames(args: {
  productName?: string;
  platformName?: string;
  platform: { platformName: string };
}) {
  const { productName, platformName, platform } = args;

  const resolvedPlatformName = platform.platformName ?? platformName;
  let finalProductName = "";
  let finalPlatformName = "";

  if (productName && resolvedPlatformName) {
    finalProductName = productName;
    finalPlatformName = resolvedPlatformName;
  } else if (productName && !resolvedPlatformName) {
    finalProductName = productName;
  } else if (!productName && resolvedPlatformName) {
    finalProductName = resolvedPlatformName;
  } else {
    // no-op in else branch if they don't pass anything in
  }

  return { productName: finalProductName, platformName: finalPlatformName };
}

export default UIShell;
