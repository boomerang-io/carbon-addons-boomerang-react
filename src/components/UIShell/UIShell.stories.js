import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { text, boolean } from '@storybook/addon-knobs/react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { BrowserRouter as Router, Link } from 'react-router-dom';
import {
  SideNav,
  SideNavLink,
  SideNavItems,
  SideNavMenu,
  SideNavMenuItem,
} from 'carbon-components-react';
import { Help24, ServiceDesk16 } from '@carbon/icons-react';
import LeftSideNav from '../LeftSideNav';
import { PRIVACY_DATA } from '../PrivacyStatement/constants';
import UIShell from './UIShell';

const mock = new MockAdapter(axios);

storiesOf('UIShell', module)
  .add('default', () => {
    mock.onGet('https://www.ibm.com/users/consents').reply(200, PRIVACY_DATA);
    return (
      <UIShell
        renderLogo={boolean('renderLogo', true)}
        appName={text('appName', 'Flow')}
        platformName={text('platformName', 'Boomerang')}
        baseServiceUrl="https://www.ibm.com"
        headerConfig={{
          features: {
            'community.enabled': boolean('community.enabled', true),
            'notifications.enabled': boolean('notifications.enabled', true),
            'support.enabled': boolean('support.enabled', true),
          },
          navigation: [
            {
              name: 'Launchpad',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
            {
              name: 'Admin',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
            {
              name: 'Docs',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
          ],
          platform: {
            name: text('platform.name', 'IBM Boomerang Platform'),
            version: text('platform.version', '5.0.0'),
            signOutUrl: 'https://ibm.com',
            communityUrl: 'https://developer.ibm.com',
          },
          platformMessage: {
            kind: text('platformMessage.kind', 'info'),
            message: text('platformMessage.message', 'Message Goes Here'),
            title: text('platformMessage.title', 'Testing Platform Title'),
          },
        }}
        onTutorialClick={action('Tutorial')}
        skipToContentProps={{ href: '#id' }}
        user={{
          name: 'test user',
          email: 'test.user@ibm.com',

          hasConsented: true,
          status: 'active',
        }}
      />
    );
  })
  .add('with Carbon sidenav and React Router ', () => {
    mock.onGet('https://www.ibm.com/users/consents').reply(200, PRIVACY_DATA);
    return (
      <Router>
        <UIShell
          renderLogo={boolean('renderLogo', true)}
          platformName={text('platformName', 'Boomerang')}
          appName={text('appName', '')}
          baseServiceUrl="https://www.ibm.com"
          headerConfig={{
            features: {
              'community.enabled': boolean('community.enabled', true),
              'notifications.enabled': boolean('notifications.enabled', true),
              'support.enabled': boolean('support.enabled', true),
            },
            navigation: [
              {
                name: 'Launchpad',
                url: 'https://launch.boomerangplatform.net/launchpad',
              },
              {
                name: 'Admin',
                url: 'https://launch.boomerangplatform.net/launchpad',
              },
              {
                name: 'Docs',
                url: 'https://launch.boomerangplatform.net/launchpad',
              },
            ],
            platform: {
              name: text('platform.name', 'IBM Boomerang Platform'),
              version: text('platform.version', '5.0.0'),
              signOutUrl: 'https://ibm.com',
              communityUrl: 'https://developer.ibm.com',
            },
            platformMessage: {
              kind: text('platformMessage.kind', 'info'),
              message: text('platformMessage.message', 'Message Goes Here'),
              title: text('platformMessage.title', 'Testing Platform Title'),
            },
          }}
          renderSidenav={({ isOpen }) => (
            <LeftSideNav isOpen={isOpen}>
              <SideNav expanded isChildOfHeader aria-label="sidenav">
                <SideNavItems>
                  <SideNavLink element={Link} to="">
                    Link
                  </SideNavLink>
                  <SideNavLink isActive element={Link} renderIcon={ServiceDesk16} to="">
                    Active link with icon
                  </SideNavLink>
                  <SideNavLink element={Link} large>
                    Large link
                  </SideNavLink>
                  <SideNavLink isActive element={Link} renderIcon={ServiceDesk16} to="" large>
                    Large active link with icon
                  </SideNavLink>
                  <SideNavMenu title="Menu">
                    <SideNavMenuItem element={Link} to="">
                      Active menu item 1
                    </SideNavMenuItem>
                    <SideNavMenuItem href="">Menu item 2</SideNavMenuItem>
                    <SideNavMenuItem href="">Menu item 3</SideNavMenuItem>
                  </SideNavMenu>
                  <SideNavMenu renderIcon={ServiceDesk16} title="Active menu with icon">
                    <SideNavMenuItem isActive element={Link} to="">
                      Active menu item 1
                    </SideNavMenuItem>
                    <SideNavMenuItem href="">Menu item 2</SideNavMenuItem>
                    <SideNavMenuItem href="">Menu item 3</SideNavMenuItem>
                  </SideNavMenu>
                  <SideNavMenu title="Large menu" large>
                    <SideNavMenuItem element={Link} to="/">
                      Large menu item 1
                    </SideNavMenuItem>
                    <SideNavMenuItem element={Link} to="/">
                      Large menu item 2
                    </SideNavMenuItem>
                  </SideNavMenu>
                  <SideNavMenu renderIcon={ServiceDesk16} title="Large active menu with icon" large>
                    <SideNavMenuItem isActive element={Link} to="">
                      Large active menu item 1
                    </SideNavMenuItem>
                    <SideNavMenuItem href="">Large menu item 2</SideNavMenuItem>
                    <SideNavMenuItem href="">Large menu item 3</SideNavMenuItem>
                  </SideNavMenu>
                </SideNavItems>
              </SideNav>
            </LeftSideNav>
          )}
          onTutorialClick={action('Tutorial')}
          user={{
            name: 'test user',
            email: 'test.user@ibm.com',
          }}
        />
      </Router>
    );
  })
  .add('with right panel', () => {
    mock.onGet('https://www.ibm.com/users/consents').reply(200, PRIVACY_DATA);
    return (
      <UIShell
        renderLogo={boolean('renderLogo', true)}
        appName={text('appName', '')}
        platformName={text('platformName', 'Boomerang')}
        baseServiceUrl="https://www.ibm.com"
        headerConfig={{
          features: {
            'community.enabled': boolean('community.enabled', true),
            'notifications.enabled': boolean('notifications.enabled', true),
            'support.enabled': boolean('support.enabled', true),
          },
          navigation: [
            {
              name: 'Launchpad',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
            {
              name: 'Admin',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
            {
              name: 'Docs',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
          ],
          platform: {
            name: text('platform.name', 'IBM Boomerang Platform'),
            version: text('platform.version', '5.0.0'),
            signOutUrl: 'https://ibm.com',
            communityUrl: 'https://developer.ibm.com',
          },
          platformMessage: {
            kind: text('platformMessage.kind', 'info'),
            message: text('platformMessage.message', 'Message Goes Here'),
            title: text('platformMessage.title', 'Testing Platform Title'),
          },
        }}
        onTutorialClick={action('Tutorial')}
        renderRightPanel={{
          icon: <Help24 />,
          component: (
            <div
              style={{
                color: 'white',
                display: 'flex',
                justifyContent: 'center',
                marginTop: '4rem',
                width: '25rem',
              }}
            >
              Custom content behaviour
            </div>
          ),
        }}
        user={{
          name: 'test user',
          email: 'test.user@ibm.com',
          hasConsented: true,
          status: 'active',
        }}
      />
    );
  })
  .add('user not consented', () => {
    mock.onGet('https://www.ibm.com/users/consents').reply(200, PRIVACY_DATA);
    return (
      <UIShell
        renderLogo={boolean('renderLogo', true)}
        appName={text('appName', '')}
        baseServiceUrl="https://www.ibm.com"
        platformName={text('platformName', 'Boomerang')}
        headerConfig={{
          features: {
            'community.enabled': boolean('community.enabled', true),
            'notifications.enabled': boolean('notifications.enabled', true),
            'support.enabled': boolean('support.enabled', true),
          },
          navigation: [
            {
              name: 'Launchpad',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
            {
              name: 'Admin',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
            {
              name: 'Docs',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
          ],
          platform: {
            name: text('platform.name', 'IBM Boomerang Platform'),
            version: text('platform.version', '5.0.0'),
            signOutUrl: 'https://ibm.com',
            communityUrl: 'https://developer.ibm.com',
          },
          platformMessage: {
            kind: text('platformMessage.kind', 'info'),
            message: text('platformMessage.message', 'Message Goes Here'),
            title: text('platformMessage.title', 'Testing Platform Title'),
          },
        }}
        onTutorialClick={action('Tutorial')}
        user={{
          name: 'test user',
          email: 'test.user@ibm.com',

          hasConsented: false,
          status: 'active',
        }}
      />
    );
  })
  .add('user pending deletion', () => {
    mock.onGet('https://www.ibm.com/users/consents').reply(200, PRIVACY_DATA);
    return (
      <UIShell
        renderLogo={boolean('renderLogo', true)}
        platformName={text('platformName', 'Boomerang')}
        appName={text('appName', '')}
        renderLogo={boolean('renderLogo', true)}
        baseServiceUrl="https://www.ibm.com"
        headerConfig={{
          features: {
            'community.enabled': boolean('community.enabled', true),
            'notifications.enabled': boolean('notifications.enabled', true),
            'support.enabled': boolean('support.enabled', true),
          },
          navigation: [
            {
              name: 'Launchpad',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
            {
              name: 'Admin',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
            {
              name: 'Docs',
              url: 'https://launch.boomerangplatform.net/launchpad',
            },
          ],
          platform: {
            name: text('platform.name', 'IBM Boomerang Platform'),
            version: text('platform.version', '5.0.0'),
            signOutUrl: 'https://ibm.com',
            communityUrl: 'https://developer.ibm.com',
          },
          platformMessage: {
            kind: text('platformMessage.kind', 'info'),
            message: text('platformMessage.message', 'Message Goes Here'),
            title: text('platformMessage.title', 'Testing Platform Title'),
          },
        }}
        onTutorialClick={action('Tutorial')}
        user={{
          name: 'test user',
          email: 'test.user@ibm.com',

          hasConsented: false,
          status: 'pending_deletion',
        }}
      />
    );
  })
  .add('without props', () => {
    mock.onGet('https://www.ibm.com/users/consents').reply(200, PRIVACY_DATA);
    return <UIShell />;
  });
