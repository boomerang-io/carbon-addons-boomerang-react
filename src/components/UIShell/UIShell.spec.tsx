/*
IBM Confidential
694970X, 69497O0
Â© Copyright IBM Corp. 2022, 2024
*/


import React from "react";
import { expect, test, vi } from "vitest";
import { screen, render, waitForElementToBeRemoved } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { axe } from "jest-axe";
import { UIShellKitchenSink } from "./UIShell.stories";
import UIShell from "./UIShell";

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: vi.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(), // Deprecated
    removeListener: vi.fn(), // Deprecated
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
});

describe("UIShell", () => {
  test("snapshot", async () => {
    const { baseElement } = render(<UIShell productName="Boomerang" />);
    expect(baseElement).toMatchSnapshot();
  });

  test("functional", async () => {
    const user = userEvent.setup();
    render(<UIShellKitchenSink />);

    // Get name
    expect(screen.getByText("Boomerang")).toBeInTheDocument();
    expect(screen.getByText("Flow")).toBeInTheDocument();

    // Get nav links
    expect(screen.getAllByText("Launchpad").length).toBe(2);
    expect(screen.getAllByText("Admin").length).toBe(2);
    expect(screen.getAllByText("Docs").length).toBe(2);

    // Get header menus and interact with them
    const requestsMenuButton = screen.getByLabelText(/Requests menu/i);
    const notificationsMenuButton = screen.getByLabelText(/Notifications dialog/i);
    const supporttMenuButton = screen.getByLabelText(/Support menu/i);
    const profileMenuButton = screen.getByLabelText(/Profile menu/i);
    const appSwitcherButton = screen.getByLabelText(/Switcher menu/i);

    expect(requestsMenuButton).toBeInTheDocument();
    expect(notificationsMenuButton).toBeInTheDocument();
    expect(supporttMenuButton).toBeInTheDocument();
    expect(profileMenuButton).toBeInTheDocument();
    expect(appSwitcherButton).toBeInTheDocument();

    user.click(requestsMenuButton);
    expect(await screen.findByText(/17 Requests/i)).toBeInTheDocument();
    expect(await screen.findByText(/11 Requests/i)).toBeInTheDocument();

    user.click(notificationsMenuButton);
    await waitForElementToBeRemoved(() => screen.queryByText(/17 Requests/i));
    expect(await screen.findByText(/Open Notification Center/i)).toBeInTheDocument();

    user.click(supporttMenuButton);
    expect(await screen.findByRole("link", { name: /Community/i })).toBeInTheDocument();

    user.click(profileMenuButton);
    expect(await screen.findByRole("link", { name: /Rick Deckard/i })).toBeInTheDocument();
    expect(await screen.findByRole("link", { name: /Email Preferences/i })).toBeInTheDocument();
    expect(await screen.findByRole("link", { name: /App Policy/i })).toBeInTheDocument();
  });

  test("a11y", async () => {
    const { container } = render(<UIShell productName="Boomerang" />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
