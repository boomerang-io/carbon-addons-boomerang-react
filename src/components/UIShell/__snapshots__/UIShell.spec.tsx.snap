// Vitest Snapshot v1

exports[`UIShell > snapshot 1`] = `
<body>
  <div>
    <div
      class="cds--g100 cds--layer-one"
    >
      <header
        aria-label="App navigation header"
        class="cds--header"
      >
        <div>
          <button
            aria-label="Sidenav menu"
            class="cds--header__action cds--header__menu-trigger cds--header__menu-toggle cds--header__menu-toggle__hidden"
            id="header-sidenav-menu-button"
            title="Sidenav menu"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="20"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 20 20"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2 14.8H18V16H2zM2 11.2H18V12.399999999999999H2zM2 7.6H18V8.799999999999999H2zM2 4H18V5.2H2z"
              />
            </svg>
          </button>
          <div
            inert="true"
          >
            <div
              class="cds--side-nav__overlay"
            />
            <nav
              aria-label="Side navigation"
              class="cds--side-nav__navigation cds--side-nav cds--side-nav--ux cds--side-nav--hidden"
            >
              <ul
                class="cds--side-nav__items"
              />
            </nav>
          </div>
        </div>
        <a
          class="cds--header__name"
          href=""
        >
          
          Boomerang
        </a>
        <nav
          aria-label="Platform navigation"
          class="cds--header__nav"
        >
          <ul
            class="cds--header__menu-bar"
          />
        </nav>
        <div
          class="cds--header__global"
        />
        <div
          class="Toastify"
          id="cds--bmrg-header-notifications"
        />
      </header>
    </div>
  </div>
</body>
`;
